dotnet/src/Connectors/Connectors.GoogleVertexAI/Core/Gemini/Clients/GeminiChatCompletionClient.cs

```diff
@@ -0,0 +1,581 @@
+    /// execution settings that opt-in to auto-invocation of everything in the kernel, in which case the invocation of that
+    /// prompt function could advertise itself as a candidate for auto-invocation. We don't want to outright block that,
+    /// if that's something a developer has asked to do (e.g. it might be invoked with different arguments than its parent
+    /// was invoked with), but we do want to limit it. This limit is arbitrary and can be tweaked in the future and/or made
+    /// configurable should need arise.
+    /// </remarks>
+    private const int MaxInflightAutoInvokes = 5;
+
+    /// <summary>Tracking <see cref="AsyncLocal{Int32}"/> for <see cref="MaxInflightAutoInvokes"/>.</summary>
+    private static readonly AsyncLocal<int> s_inflightAutoInvokes = new();
+
+    /// <summary>
+    /// Instance of <see cref="Meter"/> for metrics.
+    /// </summary>
+    private static readonly Meter s_meter = new(s_namespace);
+
+    /// <summary>
+    /// Instance of <see cref="Counter{T}"/> to keep track of the number of prompt tokens used.
+    /// </summary>
+    private static readonly Counter<int> s_promptTokensCounter =
+        s_meter.CreateCounter<int>(
+            name: $"{s_namespace}.tokens.prompt",
+            unit: "{token}",
+            description: "Number of prompt tokens used");
+
+    /// <summary>
+    /// Instance of <see cref="Counter{T}"/> to keep track of the number of completion tokens used.
+    /// </summary>
+    private static readonly Counter<int> s_completionTokensCounter =
+        s_meter.CreateCounter<int>(
+            name: $"{s_namespace}.tokens.completion",
+            unit: "{token}",
+            description: "Number of completion tokens used");
+
+    /// <summary>
+    /// Instance of <see cref="Counter{T}"/> to keep track of the total number of tokens used.
+    /// </summary>
+    private static readonly Counter<int> s_totalTokensCounter =
+        s_meter.CreateCounter<int>(
+            name: $"{s_namespace}.tokens.total",
+            unit: "{token}",
+            description: "Number of tokens used");
+
+    /// <summary>
+    /// Represents a client for interacting with the chat completion gemini model.
+    /// </summary>
+    /// <param name="httpClient">HttpClient instance used to send HTTP requests</param>
+    /// <param name="modelId">Id of the model supporting chat completion</param>
+    /// <param name="httpRequestFactory">Request factory for gemini rest api or gemini vertex ai</param>
+    /// <param name="endpointProvider">Endpoints provider for gemini rest api or gemini vertex ai</param>
+    /// <param name="streamJsonParser">Response streaming json parser (optional)</param>
+    /// <param name="logger">Logger instance used for logging (optional)</param>
+    public GeminiChatCompletionClient(
+        HttpClient httpClient,
+        string modelId,
+        IHttpRequestFactory httpRequestFactory,
```

Comment:
As I was suggested and I agree, move those abstractions as internal logic of your specialized clients.

`EndpointProvider` and `HttpRequestFactory` should be a client internal specialization, no need for IOC of those responsibilities.